### ISSUE TITLE
Iterate based on maximum time

### ISSUE THREAD

**Issue by leonlan** [N-Wouda/ALNS#62]
Thanks for the recent release on collecting runtimes! 

What do you think of adding a feature to iterate using a maximum time value as stopping condition? It's something I come across a lot in the literature, e.g., [François et al. (2019)](https://pubsonline.informs.org/doi/abs/10.1287/trsc.2019.0909?casa_token=ZNQK33zBuyAAAAAA:8fei1j6nVOIc9UOhSrgD0Zjg6TpYZFVvM0HqXnIFrvShskZSUFYEfNbsKKyN9rzVfKb9NKlPfGLJ).

I implemented something for myself like this:
```python
def iterate(self,
            initial_solution: State,
            weight_scheme: WeightScheme,
            crit: AcceptanceCriterion,
            max_value: float = 10_000, # originally: iterations
            condition: str = 'iterations' # new
            **kwargs) -> Result:
    """
    ...
    """
    ...

    while not self._stop(max_value, criterion):
        ...

    return Result(best, stats)


def _stop(self, max_value: float, condition: str) -> bool:
    """
    Check whether the stopping condition holds in the current iteration.
    """
    if max_value < 0:
        raise ValueError("Max value must be non-negative.")

    if not hasattr(self, "_start_time"):
        self._start_time = perf_counter()

    if not hasattr(self, "_current_iteration"):
        self._current_iteration = 0

    # Update the current values
    self._current_iteration += 1

    if condition == "iterations":
        return self._current_iteration >= max_value

    elif condition == "time":
        return perf_counter() - self._start_time >= max_value

    else:
        raise ValueError(
            "Stopping condition is invalid. Choose 'time' or 'iterations'."
        )
```

The `_stop` method is implemented such that time and iterations are always tracked so that I could collect runtimes (this was before the recent release).

**Comment by N-Wouda** [comment]
I like the idea, but propose an interface change:

```python
iterate(...,
        max_value: float = 10_000,  # originally: iterations
        condition: str = 'iterations',  # new
        ...)
```

This is fairly restrictive - what if I want to use a different condition (that is not implemented by default)? Or one that does not use `max_value`? Or needs more parameters?

That's all possibly if we replace this with a `stop` argument that takes a `StoppingCriterion`. This criterion could be a callable (with a call interface similar to `AcceptanceCriterion`?) that returns `True` if we should stop iterating, and `False` otherwise.

The iteration loop in `ALNS.iterate` then looks something like:

```python
while not stop(...):
    ...
```

Analogous to the acceptance criteria, we can then provide one or two default stopping criteria. I am thinking along your lines, like `MaxIterations` and `MaxRuntime`.

**Comment by N-Wouda** [comment]
@leonlan if you are willing, I would happily review a pull request on this!

I do not mind implementing it myself either, but that will probably take a little while before I get around to it - I am quite busy with conferences until the end of next month.

**Comment by leonlan** [comment]
Alright, I have some free time left so I'll make an attempt :-). I'll submit a pull request somewhere in the end of next week.

**Comment by N-Wouda** [comment]
Perfect, I am looking forward to your PR!

**Comment by N-Wouda** [comment]
Closed by #64.
